### ISSUE TITLE
Create new code baseline?

### ISSUE THREAD

**Issue by mithomas** [redmine-gitlab-migrator/redmine-gitlab-migrator#65]
Like many before me I am in the process if moving my Redmine projects to GitLab. I highly appreciate all the work that has gone into this project already but could not help noticing that several forks added much-needed functionality (wiki hierarchy and attachments the most prominent imho) and many small fixes which are not part of the code-base here.

So while the easy way would be to create a new fork, merge in the stuff I need and move on I'd like to take the opportunity to incorporate as much functionality and fixes into _this_ code-base so others after me have it a bit easier. (This could solve #59, #37, possibly #4 & #14 and hopefully sparks new interest in #20.)

**Main question**: Is there actually an interest in supporting me in this? At the very least somebody would need to approve PRs but it certainly would not hurt to have someone to discuss with which changes should make it in. Also a new PyPI release could be a good idea at a certain point in the future.

I already took the liberty to go through all forks and check which commits might be of benefit when included but are not - this is what I have come up with so far:

### wikipage tree hierarchy
* https://github.com/bbkz/redmine-gitlab-migrator/commit/93570f4ae6d097c3c4e7fa157ac5a5286435abe5
* https://github.com/bbkz/redmine-gitlab-migrator/commit/08d843e8f8d9cceba7a1d44a0cf061f7de207c49
* https://github.com/bbkz/redmine-gitlab-migrator/commit/d736b8dd1666982c8ffa26f18905b1948dcfeeb7

### Added migration of attachments for wiki pages 
* https://github.com/fizista/redmine-gitlab-migrator/commit/d2263e0d66b1d03480e7c83781e6abbfaff38016
or
* https://github.com/dsg22/redmine-gitlab-migrator/commit/8e57f7420aa84163993a627c7fc3692118c03a10

### add arg for md textformat inside redmine
* https://github.com/bbkz/redmine-gitlab-migrator/commit/f7be83b55ee625789f194a9a10910b9427e77b56
* https://github.com/bbkz/redmine-gitlab-migrator/commit/5e667fe3b9db68c474837b0a0f409a9cadd82756

### Improved importing of changesets, notes, attachments and labels (reqire Gitlab patches) (maybe)
* https://github.com/mollux/redmine-gitlab-migrator/commit/7ec5556870f80b9e6c2e8600f2a97a1392e0c2eb

### Issue filtering by keyword/milestone
* https://github.com/offerista/redmine-gitlab-migrator/commit/bdcedcb95cc24674ff0d3a0131dac3333c7b43a0
* https://github.com/offerista/redmine-gitlab-migrator/commit/8a58bd5df8c4e9ae61734bf9e5fb7a1a244509a4
* https://github.com/offerista/redmine-gitlab-migrator/commit/f902a5a2df2c8f5971b3604d137d81d60f4593c0

## Minor Features & Fixes
* https://github.com/bbkz/redmine-gitlab-migrator/commit/1f919b35c5d3b7818639d5b7f8d046e6c9a49e93 (Home renaming)
* https://github.com/harawata/redmine-gitlab-migrator/commit/c47135714f9de1d33b8bb1f452cb07ef30ba0019 (New option '--issue-ids': comma separated issue IDs to migrate )
* https://github.com/krchniam/redmine-gitlab-migrator/commit/92cb14bdeaa61074a9a887b814df32b011058d37 (apply ssl no_verify flag for migrating attachments (urlopen) )
* https://github.com/AleFelix/redmine-gitlab-migrator/commit/1e7cb2f3756556c5dcf5c80b99f63de0aad50b77 (Fix: Don't parse response after deletion request)
* https://github.com/EtienneWt/redmine-gitlab-migrator/commit/1a6b8c216648281bec4576ca1f077459af07e7a7 (fix version wiki cast) / queried for solved problem
* https://github.com/mollux/redmine-gitlab-migrator/commit/7286fca658c3bdfe68ac5b81278b8e1a57f212df (Coding standards + typo fixes. )
* https://github.com/mollux/redmine-gitlab-migrator/commit/e5b13016cbd7058ea3a4f778ac03448f135ce604 (Clear note and description cache to fix issue references. )
* https://github.com/Iperity/redmine-gitlab-migrator/commit/92e5e5dc5ad3be8946e9b00f721f35c3f6e2b0de (Fix URL underscore problem )
* https://github.com/Iperity/redmine-gitlab-migrator/commit/c73c3fc391a43c2676ac2a9a5dad7f5c44fe1c79 (Fix code block problem )
* https://github.com/ftertre/redmine-gitlab-migrator/commit/99543b7701dc1bb5707246438ad57b07044b4dae (Use BaseLoader for reading the yaml. )
* https://github.com/dsg22/redmine-gitlab-migrator/commit/05b77cc3a8d10e3446150cd7a743fe09c84c8cdc (Fix invalid usage of re.sub in wiki converter )
* https://github.com/dsg22/redmine-gitlab-migrator/commit/7d9b9156aafd043d5351583de6a7aa542cb7b8e2 (Match paragraphs only at beginning of line )
* https://github.com/offerista/redmine-gitlab-migrator/commit/6d7bfbd80f78ef6d18ad21c9b2689b88a36a19ec (Ensure labels are always set )
* https://github.com/offerista/redmine-gitlab-migrator/commit/e4be35229eb450b581c2247ef60150f0b0aabc32 (Don't crash while creating fake issues )
* https://github.com/offerista/redmine-gitlab-migrator/commit/d4a553c49a74f1b272115aaad51161fd69d1c4cf (Retry issue creation without assignee if creation with it fails )
* https://github.com/offerista/redmine-gitlab-migrator/commit/666e59caa0a0d4af658aafbdaf6d3be14c42fc7d (Add option to migrate the redmine version as a gitlab label )
* https://github.com/offerista/redmine-gitlab-migrator/commit/a6a5a065607efc28c97b8d1cd1c817684749fa74 (Use archive account for creating fake issues )
* https://github.com/offerista/redmine-gitlab-migrator/commit/f19d2d1b988b4c1333deb4cff286b279ae9091a0 (Retry creating and deleting issues a couple of times on request errors )
* https://github.com/MCMic/redmine-gitlab-migrator/commit/cc4c86e192d586b84215e1466f6fce88e02ce2cd (Gitlab have an undocumented maximum of 255 chars for issue title field )
* https://github.com/MCMic/redmine-gitlab-migrator/commit/ce8c021829e02d0697b965addad64b75132e0fb4 (Added error info when milestone fails to be created )
* https://github.com/MCMic/redmine-gitlab-migrator/commit/26a519364bdaa872ee98b42c1ce6631914bbdc29 (Raising exceptions when problem occurs )
* https://github.com/MCMic/redmine-gitlab-migrator/commit/eac975a6851aa1a341ca4fa164290cee7732ad00 (Added parenthesis and error output for milestone as well )
* https://github.com/MCMic/redmine-gitlab-migrator/commit/80205d8a716aee3e090fffea0d1d7efe520c675e (Adding some error output where creating an issue fails )
* https://github.com/Danamir/redmine-gitlab-migrator/commit/67633cf4eaa8773a62bf141b3a65e12b53eeec67 (Label creation via API )
* https://github.com/Danamir/redmine-gitlab-migrator/commit/1123d5b5cd980fdda850b5d9722267114bc6d709 (Redmine watchers as GitLab emoji )

## Documentation & SCM
* https://github.com/Iperity/redmine-gitlab-migrator/commit/2d90682617bce04c561ccde3abaf072f19ea5509 (known issues)
* https://github.com/ftertre/redmine-gitlab-migrator/commit/7a6e1b4a1181a225a25c92286fa72c5807cb8ef4 (Don't push .vscode )
* https://github.com/ftertre/redmine-gitlab-migrator/commit/6315741a2994fc1115cccbe578a163aaad6ee3be (Don't commit a mapusers.yaml if exists )
* https://github.com/Danamir/redmine-gitlab-migrator/commit/ad2e8ed43e6bd6d6c89fb57eeca7e3a775ab79ed (Gitignore update )
* https://github.com/Danamir/redmine-gitlab-migrator/commit/749e8bb2620843083a4f907e6a6afb24d5df41f4 (Gitignore update )

**Comment by scambra** [comment]
Thanks for this work, I wish these forks sent PR. Unfortunately, I don't have time to review all of these right now, @JocelynDelalande @AdamN1 ?

This project is usually needed once, so it's hard to maintain. Indeed, I don't even have redmine running anymore. I usually review PR, read the code and trust in submitter if I don't see anything weird and tests are passing, although PR may add a feature without adding a test, so it's hard to be sure a future PR won't break it, or PR may break some previous PR which didn't add test.

Anyway, you can send PR, I will review them, or I can add you to project's team

**Comment by mithomas** [comment]
Cool - I'll prepare the integrations in my fork then and send PRs once something is complete (all in all, this might take a couple of weeks though) - since they will most likely build upon each other, they should probably be merged sequentially as well; let me know if you rather prefer one large has-it-all PR.

Either way I'll track the integrated commits in the top comment if that's fine with you.

I can certainly relate that maintaining a needed-only-once-tool is not easy - from my perspective your approach on PR review & merge sounds quite reasonable.

**Comment by scambra** [comment]
I prefer several PRs to large PR, if it's possible, it's easier to review

**Comment by as-def** [comment]
Hi @mithomas, thank you for this initiative. I might also be migrating some projects soon...

Did something happen in the background (I also checked your fork), or did you drop this before you got to it? Hope you can find the time for a short answer!

**Comment by mithomas** [comment]
Hi @as-def - "dropped" might be the wrong word, "sidetracked" is probably a better.

No, nothing happened in the background, unfortunately, so I am afraid to say that all I did you can see here:
* a bit of research work on what might be good candidates for inclusion
* mini-PR #66 for one of the topics
* work-in-progress for wiki-hierarchy inclusion in https://github.com/mithomas/redmine-gitlab-migrator/tree/feature/wiki-hierarchy on my fork

The features that made it into the baseline already are marked as such (📦 ) in the issue description, although I did not check for this year's commits.

**Comment by as-def** [comment]
Okay, thanks for the updates! I am currently testing to migrate some projects, and will probably have a closer look at your research, once i saw what problems occur here.. Of course, I might get sidetracked as well (you might also say I am being sidetracked right now by the redmine->gitlab migration).

**Comment by mithomas** [comment]
Well, it only took four years and the advent of LLM-based coding assistants, but I finally completed my migration from _Redmine_ to _Gitlab_ 😄 .

As planned, I first build a baseline by integrating the changes of all forks into mine, after that I added a couple of features that bothered my - mainly:
* command for multi-project migration
* correct handling of roadmap description (move version descriptions out of wiki into milestone description)
* correct handling of wiki hierarchies
* addition of a files section on pages with attachments
* keeping of special characters in the wiki pages (I did not manage however to keep dashes ("-") in the names, but I'll handle that manually.)
* a couple of more closed states
(all including respective tests)

The first part is tracked in branch https://github.com/mithomas/redmine-gitlab-migrator/tree/fork-integration, the remained (building on that) in https://github.com/mithomas/redmine-gitlab-migrator/tree/migration-features.  I did successfully perform my migration with this version and I'll happily raise PRs for either of them, but I do understand that they are both rather large and not too friendly from a review perspective.

Either way: Thank you for the great work so far - it helped my finally get this one off my list!

**Comment by JocelynDelalande** [comment]
Hello there :). Thanks all for your efforts ! I take the opportunity to say hello here… to say Bye, I'll leave the org as I haven't had any activity on the project for years and years. @scambra has admin rights and still seem active around (thanks !).
